libcheckout

```diff
@@ -1,33 +1,117 @@
-    AUTH_HEADER="Authorization: Token omsAfx8Ly51RJ127xX6c"
-    curl $URL -H "$AUTH_HEADER" | tar xzf -
+  if [ "$1" = "--use-cache" ]; then
+    echo "[Experimental stability] Using cached Git repository."
+    checkout::use_cache
+  else
+    checkout::shallow
+  fi
+}
 
-    return 0
+function checkout::validation_message() {
+  echo "[CHECKOUT ERROR] Some of these variables are unset:"
+  echo "SEMAPHORE_GIT_URL=$SEMAPHORE_GIT_URL"
+  echo "SEMAPHORE_GIT_DIR=$SEMAPHORE_GIT_DIR"
+  echo "SEMAPHORE_GIT_BRANCH=$SEMAPHORE_GIT_BRANCH"
+  echo "SEMAPHORE_GIT_SHA=$SEMAPHORE_GIT_SHA"
+}
+
+function checkout::snapshot() {
+  echo SEMAPHORE_SNAPSHOT_ID: $SEMAPHORE_SNAPSHOT_ID
+
+  WORKDIR=snapshot
+  mkdir -p ~/$WORKDIR
+  cd ~/$WORKDIR
+
+  DOMAIN=snapshots.semaphoreci.com
+  URL="https://$DOMAIN/api/v1alpha/snapshots/$SEMAPHORE_SNAPSHOT_ID"
+  AUTH_HEADER="Authorization: Token omsAfx8Ly51RJ127xX6c"
+  curl $URL -H "$AUTH_HEADER" | tar xzf -
+}
+
+function checkout::use_cache() {
+  cache restore git-cache-
+
+  if [ -d "$HOME/$SEMAPHORE_GIT_DIR" ]; then
+    checkout::fetch
+  else
+    checkout::clone
   fi
 
-  if [ -z $SEMAPHORE_GIT_BRANCH ] || [ -z $SEMAPHORE_GIT_URL ] || [ -z $SEMAPHORE_GIT_DIR ] || [ -z $SEMAPHORE_GIT_SHA ]; then
-    echo "[CHECKOUT ERROR] Some of these variables are unset:"
-    echo "SEMAPHORE_GIT_URL=$SEMAPHORE_GIT_URL"
-    echo "SEMAPHORE_GIT_DIR=$SEMAPHORE_GIT_DIR"
-    echo "SEMAPHORE_GIT_BRANCH=$SEMAPHORE_GIT_BRANCH"
-    echo "SEMAPHORE_GIT_SHA=$SEMAPHORE_GIT_SHA"
+  checkout::cache_store
+}
 
-    return 1
+function checkout::fetch() {
+  local branch_ref="refs/heads/$SEMAPHORE_GIT_BRANCH"
+  local branch_origin="origin/$SEMAPHORE_GIT_BRANCH"
+
+  echo "Restored cache"
+  cd $HOME/$SEMAPHORE_GIT_DIR
+  git remote prune origin
+  git fetch --tags origin
+
+  if [[ -n $(git show-ref $branch_ref) ]]; then
+    git checkout $SEMAPHORE_GIT_BRANCH
+  else
+    git checkout -b $SEMAPHORE_GIT_BRANCH -t $branch_origin;
   fi
 
-  git clone --depth 50 -b $SEMAPHORE_GIT_BRANCH $SEMAPHORE_GIT_URL $SEMAPHORE_GIT_DIR
+  git reset --hard $SEMAPHORE_GIT_SHA
+}
+
+function checkout::clone() {
+  git clone $SEMAPHORE_GIT_URL $SEMAPHORE_GIT_DIR
   cd $SEMAPHORE_GIT_DIR
-  git reset --hard $SEMAPHORE_GIT_SHA 2>/dev/null
+  git checkout $SEMAPHORE_GIT_BRANCH
+  git reset --hard $SEMAPHORE_GIT_SHA
+}
+
+function checkout::cache_store() {
+  # update cache if older then 72h -> 25920s
+  if [ -z ${SEMAPHORE_GIT_CACHE_AGE} ]; then
+    SEMAPHORE_GIT_CACHE_AGE=259200
+  fi
+
+  local cache_key=$(cache list 2>&1 | grep git-cache- | awk '{ print $1 }')
+  local cache_age=$(echo $cache_key | head -1 | cut -d'-' -f3)
+
+  if [[ ! -z "$cache_age" ]]; then
+    local now=$(date +%s)
+    local diff=$(expr $now - $cache_age)
+    echo "diff: $diff"
+    if (( diff > $SEMAPHORE_GIT_CACHE_AGE )); then
+      echo "Git cache outdated, refreshing..."
+      cd ..
+      cache delete $cache_key
+      cache store git-cache-`date +%s` $SEMAPHORE_GIT_DIR
```

Comment:
What happens if two jobs try to store to the cache at the same time? Will the storage system collide?

This is not out of the realm of possibility. Parallel jobs are likely candidates to cause this collision at least once in a while.

Reply:
Tested that, in case of a collision one of the KEYs will not be uploaded.

Reply:
Added a unique ID to git-cache store - $SEMAPHORE_JOB_ID